torch/csrc/jit/import.cpp

```diff
@@ -125,62 +125,84 @@ script::Module ScriptModuleDeserializer::deserialize(
   AT_ASSERTM(
       model_def.ParseFromString(binary_string),
       "JSON transcoder produced invalid protobuf output.");
-  device_ = device;
-  proto_version_ = model_def.proto_version();
+  auto proto_version = model_def.proto_version();
+  export_prefix_ = "libs/";
+
+  LEGACY_loadTensorTable(&model_def);
+  AT_ASSERT(proto_version < 6);
+  if (proto_version == 2) {
+    const auto& list =
+        LEGACY_loadPickleArchive("attributes.pkl").toGenericList();
+    LEGACY_pickled_ivalues_.insert(
+        LEGACY_pickled_ivalues_.end(), list.begin(), list.end());
+  } else if (proto_version >= 3) {
+    LEGACY_pickled_ivalues_ =
+        LEGACY_loadPickleArchive("attributes.pkl").toTuple()->elements();
+  }
+  LEGACY_moduleStack_.push_back("__torch__");
+  const auto& module_def = model_def.main_module();
+  return LEGACY_convertModule(module_def);
+}
+
+IValue ScriptModuleDeserializer::readArchive(const std::string& archive_name) {
+  std::stringstream picklename;
+  picklename << archive_name << ".pkl";
+  at::DataPtr pickle_ptr;
+  size_t pickle_size;
+  std::tie(pickle_ptr, pickle_size) = reader_->getRecord(picklename.str());
+
+  size_t bytes_read = 0;
+  auto data = reinterpret_cast<const char*>(pickle_ptr.get());
+  auto reader = [&](char* buffer, size_t len) {
+    if (bytes_read + len > pickle_size) {
+      return false;
+    }
+    // Copy len bytes into buffer
+    const char* start = data + bytes_read;
+    std::memcpy(buffer, start, len);
+    bytes_read += len;
+    return true;
+  };
+
+  auto class_resolver = [&](const c10::QualifiedName& qn) {
+    importCallback(qn.prefix());
+    return c10::StrongTypePtr(
+        compilation_unit_, compilation_unit_->get_class(qn));
+  };
+  auto read_record = [&](const std::string& name) {
+    std::stringstream ss;
+    ss << archive_name << "/" << name;
+    return std::get<0>(reader_->getRecord(ss.str()));
+  };
+  Unpickler unpickler(
```

Comment:
Same here, this can be `jit::unpickle`

Reply:
user-exposed pickle functions do not take the `read_record` or `device_` arguments needed here to load the tensors.